File: src/flutter/papernestextension/lib/src/controls/papernest_textfield.dart
import 'dart:async';

import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

class PaperNestTextFieldControl extends StatefulWidget {
  final Control control;

  PaperNestTextFieldControl({Key? key, required this.control})
      : super(key: key ?? ValueKey("control_${control.id}"));

  @override
  State<PaperNestTextFieldControl> createState() => _PaperNestTextFieldControlState();
}

class _PaperNestTextFieldControlState extends State<PaperNestTextFieldControl> {
  String _value = "";
  bool _revealPassword = false;
  bool _focused = false;
  late TextEditingController _controller;
  late final FocusNode _focusNode;
  late final FocusNode _shiftEnterfocusNode;
  String? _lastFocusValue;
  String? _lastBlurValue;
  TextSelection? _selection;
  Timer? _searchDebounce;

  KeyEventResult _handleTextFieldKeyEvent(KeyEvent event,
      {required bool submitOnEnter}) {
    if (event is KeyDownEvent &&
        event.logicalKey == LogicalKeyboardKey.escape) {
      final searchMode = widget.control.getBool("search_mode", false)!;
      final clearOnEscape = widget.control.getBool("clear_on_escape", true)!;
      final blurOnEmptyEscape =
          widget.control.getBool("blur_on_empty_escape", true)!;

      widget.control.triggerEvent("escape", _controller.text);

      if (searchMode && clearOnEscape && _controller.text.isNotEmpty) {
        _clearSearch();
      } else if (blurOnEmptyEscape) {
        _focusNode.unfocus();
        _shiftEnterfocusNode.unfocus();
      }

      return KeyEventResult.handled;
    }

    // ignore up/down arrow keys if flag is set
    if ((event is KeyDownEvent || event is KeyRepeatEvent) &&
        widget.control.getBool("ignore_up_down_keys", false)! &&
        (event.logicalKey == LogicalKeyboardKey.arrowUp ||
            event.logicalKey == LogicalKeyboardKey.arrowDown)) {
      return KeyEventResult.handled;
    }

    // submit on Enter if flag is set and shift is not pressed
    if (submitOnEnter &&
        event is KeyDownEvent &&
        !HardwareKeyboard.instance.isShiftPressed &&
        (event.logicalKey == LogicalKeyboardKey.enter ||
            event.logicalKey == LogicalKeyboardKey.numpadEnter)) {
      widget.control.triggerEvent("submit");
      return KeyEventResult.handled;
    }

    // let the system handle other key events
    return KeyEventResult.ignored;
  }

  @override
  void initState() {
    super.initState();
    _controller = TextEditingController();
    _controller.addListener(_handleControllerChange);
    _shiftEnterfocusNode = FocusNode(
      onKeyEvent: (FocusNode node, KeyEvent event) =>
          _handleTextFieldKeyEvent(event, submitOnEnter: true),
    );
    _shiftEnterfocusNode.addListener(_onShiftEnterFocusChange);
    _focusNode = FocusNode(
      onKeyEvent: (FocusNode node, KeyEvent event) =>
          _handleTextFieldKeyEvent(event, submitOnEnter: false),
    );
    _focusNode.addListener(_onFocusChange);
    widget.control.addInvokeMethodListener(_invokeMethod);
  }

  @override
  void dispose() {
    _searchDebounce?.cancel();
    _controller.removeListener(_handleControllerChange);
    _controller.dispose();
    _shiftEnterfocusNode.removeListener(_onShiftEnterFocusChange);
    _shiftEnterfocusNode.dispose();
    _focusNode.removeListener(_onFocusChange);
    widget.control.removeInvokeMethodListener(_invokeMethod);
    _focusNode.dispose();
    super.dispose();
  }

  Future<dynamic> _invokeMethod(String name, dynamic args) async {
    debugPrint("TextField.$name($args)");
    switch (name) {
      case "focus":
        _focusNode.requestFocus();
      default:
        throw Exception("Unknown TextField method: $name");
    }
  }

  void _selectAllIfRequested(bool hasFocus) {
    if (!hasFocus ||
        !widget.control.getBool("select_all_on_focus", false)! ||
        _controller.text.isEmpty) {
      return;
    }

    WidgetsBinding.instance.addPostFrameCallback((_) {
      if (!mounted) return;
      _controller.selection = TextSelection(
        baseOffset: 0,
        extentOffset: _controller.text.length,
      );
    });
  }

  void _onShiftEnterFocusChange() {
    _focused = _shiftEnterfocusNode.hasFocus;
    _selectAllIfRequested(_focused);
    widget.control
        .triggerEvent(_shiftEnterfocusNode.hasFocus ? "focus" : "blur");
  }

  void _onFocusChange() {
    _focused = _focusNode.hasFocus;
    _selectAllIfRequested(_focused);
    widget.control.triggerEvent(_focusNode.hasFocus ? "focus" : "blur");
  }

  void _clearSearch() {
    _searchDebounce?.cancel();
    _controller.clear();
    _value = "";
    widget.control.updateProperties({"value": ""});
    widget.control.triggerEvent("clear", "");
    widget.control.triggerEvent("search", "");
    if (mounted) {
      setState(() {});
    }
  }

  void _handleControllerChange() {
    final selection = _controller.selection;
    if (_selection == selection) return;

    _selection = selection;

    if (!selection.isValid ||
        !widget.control.hasEventHandler("selection_change")) {
      return;
    }

    widget.control.updateProperties({"selection": selection.toMap()});
    widget.control.triggerEvent("selection_change", {
      "selected_text":
          _controller.text.substring(selection.start, selection.end),
      "selection": selection.toMap()
    });
  }

  @override
  Widget build(BuildContext context) {
    debugPrint("TextField build: ${widget.control.id}");

    bool autofocus = widget.control.getBool("autofocus", false)!;

    String value = widget.control.getString("value", "")!;
    if (_value != value) {
      _value = value;
      _controller.value = TextEditingValue(
        text: value,
        // preserve cursor position at the end
        selection: TextSelection.collapsed(offset: value.length),
      );
      _selection = _controller.selection;
    }

    var selection = widget.control.getTextSelection("selection",
        minOffset: 0, maxOffset: _controller.text.length);
    if (selection != null && selection != _controller.selection) {
      _controller.selection = selection;
      _selection = selection;
    }

    var shiftEnter = widget.control.getBool("shift_enter", false)!;
    var multiline = widget.control.getBool("multiline", false)! || shiftEnter;
    var minLines = widget.control.getInt("min_lines", 1)!;
    var maxLines = widget.control.getInt("max_lines", multiline ? null : 1);

    var password = widget.control.getBool("password", false)!;
    var canRevealPassword =
        widget.control.getBool("can_reveal_password", false)!;
    var cursorColor = widget.control.getColor("cursor_color", context);
    var selectionColor = widget.control.getColor("selection_color", context);
    var textSize = widget.control.getDouble("text_size");
    var color = widget.control.getColor("color", context);
    var focusedColor = widget.control.getColor("focused_color", context);
    var textStyle = widget.control
        .getTextStyle("text_style", Theme.of(context), const TextStyle())!;
    if (textSize != null || color != null || focusedColor != null) {
      textStyle = textStyle.copyWith(
          fontSize: textSize, color: _focused ? focusedColor ?? color : color);
    }

    TextCapitalization textCapitalization = widget.control
        .getTextCapitalization("capitalization", TextCapitalization.none)!;

    FilteringTextInputFormatter? inputFilter =
        widget.control.getTextInputFormatter("input_filter");

    List<TextInputFormatter>? inputFormatters = [];
    // add non-null input formatters
    if (inputFilter != null) {
      inputFormatters.add(inputFilter);
    }
    if (textCapitalization != TextCapitalization.none) {
      inputFormatters.add(TextCapitalizationFormatter(textCapitalization));
    }

    final searchMode = widget.control.getBool("search_mode", false)!;
    final clearButton = widget.control.getBool("clear_button", true)!;
    final searching = widget.control.getBool("searching", false)!;
    final showRefreshAction =
        widget.control.getBool("show_refresh_action", false)!;

    final fieldState =
        widget.control.getString("state", "normal")!.toLowerCase();
    final stateMessage = widget.control.getString("state_message");
    final stateColor = switch (fieldState) {
      "success" => Colors.green,
      "warning" => Colors.orange,
      "error" => Theme.of(context).colorScheme.error,
      _ => null,
    };
    final stateIcon = switch (fieldState) {
      "success" => Icons.check_circle_outline,
      "warning" => Icons.warning_amber_rounded,
      "error" => Icons.error_outline,
      _ => null,
    };

    Widget? suffixIcon;
    if (password && canRevealPassword) {
      suffixIcon = IconButton(
        icon: Icon(
          _revealPassword ? Icons.visibility_off : Icons.visibility,
        ),
        tooltip: _revealPassword ? "Masquer" : "Afficher",
        onPressed: () {
          setState(() {
            _revealPassword = !_revealPassword;
          });
        },
      );
    } else if ((searchMode &&
            (searching || showRefreshAction ||
                (clearButton && _value.isNotEmpty))) ||
        stateIcon != null) {
      suffixIcon = Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          if (searching)
            const Padding(
              padding: EdgeInsets.symmetric(horizontal: 12),
              child: SizedBox.square(
                dimension: 16,
                child: CircularProgressIndicator(strokeWidth: 2),
              ),
            ),
          if (showRefreshAction)
            IconButton(
              icon: const Icon(Icons.refresh),
              tooltip: widget.control
                  .getString("refresh_action_tooltip", "Actualiser"),
              onPressed:
                  widget.control.getBool("refresh_action_disabled", false)!
                      ? null
                      : () => widget.control.triggerEvent("refresh_action"),
            ),
          if (stateIcon != null && stateColor != null)
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 10),
              child: Icon(stateIcon, color: stateColor, size: 20),
            ),
          if (clearButton && _value.isNotEmpty)
            IconButton(
              icon: const Icon(Icons.close),
              tooltip: "Effacer la recherche",
              onPressed: _clearSearch,
            ),
        ],
      );
    }

    var textVerticalAlign = widget.control.getDouble("text_vertical_align");

    FocusNode focusNode = shiftEnter ? _shiftEnterfocusNode : _focusNode;
    var focusValue = widget.control.getString("focus");
    var blurValue = widget.control.getString("blur");
    if (focusValue != null && focusValue != _lastFocusValue) {
      _lastFocusValue = focusValue;
      focusNode.requestFocus();
    }
    if (blurValue != null && blurValue != _lastBlurValue) {
      _lastBlurValue = blurValue;
      _focusNode.unfocus();
    }

    var fitParentSize = widget.control.getBool("fit_parent_size", false)!;
    var maxLength = widget.control.getInt("max_length");

    InputBorder? withStateBorder(InputBorder? border, double width) {
      if (border == null || stateColor == null) return border;
      return border.copyWith(
        borderSide: BorderSide(color: stateColor, width: width),
      );
    }

    InputBorder? resolveFocusedBorder(InputDecoration decoration) {
      if (decoration.focusedBorder != null) {
        return decoration.focusedBorder;
      }

      final baseBorder = decoration.enabledBorder ?? decoration.border;
      if (baseBorder == null || baseBorder == InputBorder.none) {
        return baseBorder;
      }

      return baseBorder.copyWith(
        borderSide: BorderSide(
          color: Theme.of(context).colorScheme.primary,
          width: widget.control.getDouble("focused_border_width", 2.0)!,
        ),
      );
    }

    var decoration = buildInputDecoration(
      context,
      widget.control,
      customSuffix: suffixIcon,
      valueLength: _value.length,
      maxLength: maxLength,
      focused: _focused,
    ).copyWith(
      prefixIcon: searchMode ? const Icon(Icons.search) : null,
    );

    final focusedBorder = resolveFocusedBorder(decoration);

    if (fieldState == "error") {
      decoration = decoration.copyWith(
        errorText: stateMessage,
        enabledBorder: withStateBorder(decoration.enabledBorder, 1),
        errorBorder: withStateBorder(
          decoration.errorBorder ?? decoration.enabledBorder,
          1,
        ),
        focusedBorder: focusedBorder,
        focusedErrorBorder: focusedBorder,
      );
    } else if (stateColor != null) {
      decoration = decoration.copyWith(
        helperText: stateMessage,
        helperStyle: decoration.helperStyle?.copyWith(color: stateColor) ??
            TextStyle(color: stateColor),
        enabledBorder: withStateBorder(decoration.enabledBorder, 1),
        focusedBorder: focusedBorder,
      );
    } else if (stateMessage != null) {
      decoration = decoration.copyWith(helperText: stateMessage);
    }

    Widget textField = TextFormField(
        style: textStyle,
        autofocus: autofocus,
        enabled: !widget.control.disabled,
        onFieldSubmitted: !multiline
            ? (value) {
                _searchDebounce?.cancel();
                widget.control.triggerEvent("submit", value);
                if (searchMode) {
                  widget.control.triggerEvent("search", value);
                }
              }
            : null,
        decoration: decoration,
        showCursor: widget.control.getBool("show_cursor"),
        textAlignVertical: textVerticalAlign != null
            ? TextAlignVertical(y: textVerticalAlign)
            : null,
        cursorHeight: widget.control.getDouble("cursor_height"),
        cursorWidth: widget.control.getDouble("cursor_width", 2.0)!,
        cursorRadius: widget.control.getRadius("cursor_radius"),
        keyboardType: multiline
            ? TextInputType.multiline
            : widget.control
                .getTextInputType("keyboard_type", TextInputType.text)!,
        autocorrect: widget.control.getBool("autocorrect", true)!,
        enableSuggestions: widget.control.getBool("enable_suggestions", true)!,
        smartDashesType: widget.control.getBool("smart_dashes_type", true)!
            ? SmartDashesType.enabled
            : SmartDashesType.disabled,
        smartQuotesType: widget.control.getBool("smart_quotes_type", true)!
            ? SmartQuotesType.enabled
            : SmartQuotesType.disabled,
        textAlign: widget.control.getTextAlign("text_align", TextAlign.start)!,
        minLines: fitParentSize ? null : minLines,
        maxLines: fitParentSize ? null : maxLines,
        maxLength: maxLength,
        readOnly: widget.control.getBool("read_only", false)!,
        inputFormatters: inputFormatters.isNotEmpty ? inputFormatters : null,
        obscureText: password && !_revealPassword,
        controller: _controller,
        focusNode: focusNode,
        autofillHints: widget.control.getAutofillHints("autofill_hints"),
        expands: fitParentSize,
        enableInteractiveSelection:
            widget.control.getBool("enable_interactive_selection"),
        canRequestFocus: widget.control.getBool("can_request_focus", true)!,
        clipBehavior:
            widget.control.getClipBehavior("clip_behavior", Clip.hardEdge)!,
        cursorColor: cursorColor,
        ignorePointers: widget.control.getBool("ignore_pointers"),
        cursorErrorColor:
            widget.control.getColor("cursor_error_color", context),
        stylusHandwritingEnabled:
            widget.control.getBool("enable_stylus_handwriting", true)!,
        scrollPadding: widget.control
            .getPadding("scroll_padding", const EdgeInsets.all(20.0))!,
        keyboardAppearance: widget.control.getBrightness("keyboard_brightness"),
        enableIMEPersonalizedLearning:
            widget.control.getBool("enable_ime_personalized_learning", true)!,
        obscuringCharacter:
            widget.control.getString("obscuring_character", '•')!,
        mouseCursor: widget.control.getMouseCursor("mouse_cursor"),
        cursorOpacityAnimates: widget.control.getBool("animate_cursor_opacity",
            Theme.of(context).platform == TargetPlatform.iOS)!,
        onTapAlwaysCalled: widget.control.getBool("always_call_on_tap", false)!,
        strutStyle: widget.control.getStrutStyle("strut_style"),
        onTap: () {
          widget.control.triggerEvent("click");
        },
        onTapOutside: widget.control.hasEventHandler("tap_outside")
            ? (PointerDownEvent? event) {
                widget.control.triggerEvent("tap_outside");
              }
            : null,
        onChanged: (String value) {
          final visibilityChanged = searchMode &&
              clearButton &&
              ((_value.isEmpty && value.isNotEmpty) ||
                  (_value.isNotEmpty && value.isEmpty));

          _value = value;
          widget.control.updateProperties({"value": value});

          if (visibilityChanged) {
            setState(() {});
          }

          if (widget.control.hasEventHandler("change")) {
            widget.control.triggerEvent("change", value);
          }

          if (searchMode && widget.control.hasEventHandler("search")) {
            _searchDebounce?.cancel();
            final debounceMs = widget.control.getInt("debounce_ms", 300)!;
            if (debounceMs == 0) {
              widget.control.triggerEvent("search", value);
            } else {
              _searchDebounce = Timer(Duration(milliseconds: debounceMs), () {
                if (mounted) {
                  widget.control.triggerEvent("search", _value);
                }
              });
            }
          }
        });

    if (cursorColor != null || selectionColor != null) {
      textField = TextSelectionTheme(
          data: TextSelectionTheme.of(context).copyWith(
              cursorColor: cursorColor, selectionColor: selectionColor),
          child: textField);
    }

    // linux workaround for https://github.com/flet-dev/flet/issues/3934
    textField =
        isLinuxDesktop() ? ExcludeSemantics(child: textField) : textField;

    if (widget.control.getExpand("expand", 0)! > 0) {
      return LayoutControl(control: widget.control, child: textField);
    } else {
      double? width = widget.control.getDouble("width");

      return LayoutControl(
        control: widget.control,
        child: width == null
            ? ConstrainedBox(
                constraints: const BoxConstraints.tightFor(width: 300),
                child: textField)
            : textField,
      );
    }
  }
}

File: src/papernestextension/controls/material/papernest_textfield.py
from enum import Enum
from typing import Annotated, Optional, Union

from flet.controls.adaptive_control import AdaptiveControl
from flet.controls.base_control import BaseControl, control, value
from flet.controls.control_event import ControlEventHandler, EventHandler
from flet.controls.core.autofill_group import AutofillHint
from flet.controls.core.text import TextSelection, TextSelectionChangeEvent
from flet.controls.material.form_field_control import FormFieldControl
from flet.controls.padding import PaddingValue
from flet.controls.text_style import StrutStyle
from flet.controls.types import (
    Brightness,
    ClipBehavior,
    ColorValue,
    MouseCursor,
    Number,
    TextAlign,
)
from flet.utils.validation import V

from papernestextension.controls.control_state import ValidationState

__all__ = [
    "InputFilter",
    "KeyboardType",
    "NumbersOnlyInputFilter",
    "PaperNestTextField",
    "PaperNestTextFieldState",
    "TextCapitalization",
    "TextOnlyInputFilter",
]


class KeyboardType(Enum):
    NONE = "none"
    TEXT = "text"
    MULTILINE = "multiline"
    NUMBER = "number"
    PHONE = "phone"
    DATETIME = "datetime"
    EMAIL = "email"
    URL = "url"
    VISIBLE_PASSWORD = "visiblePassword"
    NAME = "name"
    STREET_ADDRESS = "streetAddress"
    WEB_SEARCH = "webSearch"
    TWITTER = "twitter"


PaperNestTextFieldState = ValidationState


class TextCapitalization(Enum):
    CHARACTERS = "characters"
    WORDS = "words"
    SENTENCES = "sentences"
    NONE = "none"


@value
class InputFilter:
    regex_string: str
    allow: bool = True
    replacement_string: str = ""
    multiline: bool = False
    case_sensitive: bool = True
    unicode: bool = False
    dot_all: bool = False


class NumbersOnlyInputFilter(InputFilter):
    def __init__(self):
        super().__init__(
            regex_string=r"^[0-9]*$",
            allow=True,
            replacement_string="",
        )


class TextOnlyInputFilter(InputFilter):
    def __init__(self):
        super().__init__(
            regex_string=r"^[a-zA-Z]*$",
            allow=True,
            replacement_string="",
        )


@control("PaperNestTextField")
class PaperNestTextField(FormFieldControl, AdaptiveControl):
    value: str = ""
    selection: Optional[TextSelection] = None
    keyboard_type: KeyboardType = KeyboardType.TEXT
    multiline: bool = False
    min_lines: Annotated[Optional[int], V.gt(0), V.le_field("max_lines")] = None
    max_lines: Annotated[Optional[int], V.gt(0), V.ge_field("min_lines")] = None
    max_length: Annotated[
        Optional[int],
        V.or_(
            V.gt(0),
            V.eq(-1),
            message="max_length must be either strictly greater than 0 or equal to -1",
        ),
    ] = None
    password: bool = False
    can_reveal_password: bool = False
    read_only: bool = False
    shift_enter: bool = False
    ignore_up_down_keys: bool = False
    text_align: Optional[TextAlign] = None
    autofocus: bool = False
    capitalization: Optional[TextCapitalization] = None
    autocorrect: bool = True
    enable_suggestions: bool = True
    smart_dashes_type: bool = True
    smart_quotes_type: bool = True
    show_cursor: bool = True
    cursor_color: Optional[ColorValue] = None
    cursor_error_color: Optional[ColorValue] = None
    cursor_width: Number = 2.0
    cursor_height: Optional[Number] = None
    cursor_radius: Optional[Number] = None
    selection_color: Optional[ColorValue] = None
    input_filter: Optional[InputFilter] = None
    obscuring_character: str = "•"
    enable_interactive_selection: bool = True
    enable_ime_personalized_learning: bool = True
    can_request_focus: bool = True
    ignore_pointers: bool = False
    enable_stylus_handwriting: bool = True
    animate_cursor_opacity: Optional[bool] = None
    always_call_on_tap: bool = False
    scroll_padding: PaddingValue = 20
    clip_behavior: ClipBehavior = ClipBehavior.HARD_EDGE
    keyboard_brightness: Optional[Brightness] = None
    mouse_cursor: Optional[MouseCursor] = None
    strut_style: Optional[StrutStyle] = None
    autofill_hints: Optional[Union[AutofillHint, list[AutofillHint]]] = None
    search_mode: bool = False
    clear_button: bool = True
    debounce_ms: Annotated[int, V.ge(0)] = 300
    searching: bool = False
    show_refresh_action: bool = False
    refresh_action_disabled: bool = False
    refresh_action_tooltip: str = "Actualiser"
    state: PaperNestTextFieldState = PaperNestTextFieldState.NORMAL
    state_message: Optional[str] = None
    select_all_on_focus: bool = False
    clear_on_escape: bool = True
    blur_on_empty_escape: bool = True

    on_change: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_selection_change: Optional[EventHandler[TextSelectionChangeEvent["PaperNestTextField"]]] = None
    on_click: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_submit: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_focus: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_blur: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_tap_outside: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_search: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_clear: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_escape: Optional[ControlEventHandler["PaperNestTextField"]] = None
    on_refresh_action: Optional[ControlEventHandler["PaperNestTextField"]] = None

    def _migrate_state(self, other: BaseControl):
        super()._migrate_state(other)
        if (
            isinstance(other, PaperNestTextField)
            and self.value is None
            and self.value != other.value
        ):
            self.value = other.value

    def before_update(self):
        super().before_update()
        if (
            self.bgcolor is not None
            or self.fill_color is not None
            or self.hover_color is not None
            or self.focused_color is not None
        ) and self.filled is None:
            self.filled = True

        if self.search_mode:
            self.keyboard_type = KeyboardType.WEB_SEARCH
            self.multiline = False
            self.shift_enter = False
